File: Lucy/src/2025-8-W02/127_word_ladder.ts
function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
  if (!wordList.includes(endWord)) return 0;

  const queue: [string, number][] = [[beginWord, 1]];
  const visited = new Set<string>();
  visited.add(beginWord);

  const wordListSet = new Set<string>(wordList);

  const getValidWords = (currentStr: string): string[] => {
    const results: string[] = [];

    const alphabets = "abcdefghijklmnopqrstuvwxyz";

    for (const alphabet of alphabets) {
      for (let i = 0; i < currentStr.length; i++) {
        const transformStr = currentStr.substring(0, i) + alphabet + currentStr.substring(i + 1);

        if (!visited.has(transformStr) && wordListSet.has(transformStr)) {
          results.push(transformStr);
        }
      }
    }

    return results;
  };

  while (queue.length > 0) {
    const [currentStr, transformCount] = queue.shift()!;
    if (currentStr === endWord) return transformCount;

    const validWords = getValidWords(currentStr);
    for (const word of validWords) {
      visited.add(word);
      queue.push([word, transformCount + 1]);
    }
  }

  return 0;
}

File: Seori/127. Word Ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        n = len(wordList[0])
        wordList = set(wordList)
        word_set = [set() for _ in range(n)]
        for word in wordList:
            for i in range(n):
                word_set[i].add(word[i])

        visited = set(beginWord)
        queue = [(beginWord, 1)]
        while queue:
            nowWord, count = queue.pop(0)
            if nowWord == endWord:
                return count

            for i in range(n):
                for trans_letter in word_set[i]:
                    if nowWord[i] == trans_letter: continue
                    nextWord = nowWord[:i] + trans_letter + nowWord[i+1:]
                    if nextWord in wordList and nextWord not in visited:
                        queue.append((nextWord, count+1))
                        visited.add(nextWord)
            

        return 0